README
libutap, version 0.94
Mini HOWTO
Gerd Behrmann
Marius Mikucionis


1. What is libutap
2. Compiling
3. Simple use
4. Architecture
5. Contact



1. What is libutap

libutap is the Uppaal Timed Automata Parser. Uppaal is an integrated
tool environment for designing, simulating and verifying real time
systems modeled as extended timed automata. libutap is the parser of
Uppaal.

libutap has the ability to parse and type check Uppaal models in any
of the three file formats supported by Uppaal.

libutap is licensed under the LGPL.


2. Compiling

libutap uses automake and autoconf to make compilation on various
platforms easy. You will need gcc 3.3 or newer, GNU make,
libxml2 from http://www.xmlsoft.org (at least version 2.6.10),
libsbml from http://www.sbml.org.
Simply run to compile libutap and use it:

./configure
make

Run the following to install libutap:

make install

Read the INSTALL file in the distribution for more information on
how to compile.

NOTE: if you want to regenerate automatically created files like the
configure script and Makefile, use autoreconf -i. Do *not* run
autoconf directly, it will mess up your Makefile. You need autoconf
version 2.53 or newer and automake.


3. Simple use

There are two ways one can use the library. In its simplest form, one
calls one of the top level parsing functions defined in utap/utap.h,
e.g. example.cpp:

#include <stdio.h>
#include "utap/utap.h"

int main()
{
  UTAP::TimedAutomataSystem system;
  parseXMLFile("myfile.xml", &system, true);
}

The first argument is the file to read. The second is the output of
the parser and the third is a flag indicating whether we want to use
the new or the old syntax (the old syntax is the one used in Uppaal
3.4, the new is the one that will be used since Uppaal 3.6).

After the call to parseXMLFile, one can access the network of timed
automata in the system variable. Take a look at utap/system.h to see
what kind of structures you can access.
Distribution also includes pretty.cpp for pretty-printing a model files.
See also doxygen API documentation in doc/api/index.html.

Use the following command to compile the example:

g++ example.cpp -o example -lutap -lxml2

If UTAP was configured with --prefix=$MYPATH to install in custom location
then use the following to compile:

g++ -I$MYPATH/include example.cpp -o example -L$MYPATH/lib -lutap -lxml2


4. Architecture

The following ASCII figure shows the initial information flow through
the library.

 +----\
 |    |\
 |    +-\
 |      |   --> libxml2 --[SAX-interface]--> "xmlreader.cc"
 | .xml |                                         |
 |      |                                         |
 +------+                                         |
                                                  |
 +----\                                           |
 |    |\                                          |
 |    +-\                                         v
 | .ta  |                            +--------------------------+
 | .xta |   -----------------------> | bison parser (parser.yy) |
 |      |	                     +--------------------------+
 +------+                                           |
                                                    |
 +------+                                           |
 |      |                                           |
 | TAS  |                                           |
 |      |      +---------------+                    |
 |      | <--- | SystemBuilder |<--[ParserBuilder]--/
 |      |      +---------------+
 +------+


The BNF implemented by the bison generated parser is split into
several sections and to some extend duplicated for the old and new
syntax. It can read .ta and .xta files directly. XML files are read
using the libxml2 library. Each text-block in the XML file is then
parser wrt. the BNF for that block (see xmlreader.cpp). For this, the
bison generated parser is reused.

The parser calls methods in the abstract ParserBuilder class. The
methods in the ParserBuilder class are implemented by the
SystemBuilder class which is a subclass of ExpressionBuilder which is
a subclass of AbstractBuilder. The SystemBuilder writes the model to
an instance of the TimedAutomataSystem (TAS) class.

The design abstracts the differences between the .xml, .xta and .ta
input formats and also hides the differences between the 3.4 and
3.6/4.0 formats from any implementation of the ParserBuilder interface
(for equivalent input the parser will call the same methods in the
ParserBuilder class).


A TAS object represents the templates, variables, locations,
transitions and processes of a model. Symbols are represented by
symbol_t objects (see the API documentation). A symbol is a name (a
string) with a type. The type is represented by a type_t
object. Symbols are grouped into frames (represented by frame_t
objects). Frames are used to represent scopes and other collections of
symbols such as records or parameters of templates and functions.

All expressions are represented using a tree structure where the
leaves represent values or variables and the inner nodes represent
operations. Each node is referenced using an expression_t object.

5. Contact

If you have questions or comments to libutap, you can contact me on
the following email address: marius@cs.aau.dk. For a list of known
bugs or for reporting new bugs take a look at the UPPAAL bugzilla
installation at http://www.uppaal.org/.


NEWS
New in 0.93
This is a snapshot of UTAP as used in prerelease of UPPAAL 4.1.20.
Notable changes:
- Fixed tracer for newer engine trace format.
- Fixed pretty printer handling of embedded queries.


New in 0.91

This is a snapshot of UTAP in the latest development version of 
UPPAAL. Highlights of this release include:

- Bugs fixed in this release: 322, 336, 348, 351, 352, 353, 
  358, 364, 367, 370, 372, 383.
- CTL parsing has been enhanced to allow nesting of formulas.
- TIGA support was added.
- Improved compilation on Windows.
- Syntax check utility was renamed to 'syntaxcheck'
- Type checker was cleaned up
- IO Interface utility renamed to 'taflow'
- Improved pretty printer
- Refactored SystemBuilder into SystemBuilder and StatementBuilder


New in 0.90

This is the version of UTAP used in UPPAAL 3.6 Beta 3.  The tracer
utility no longer uses libutap. Instead it parses the UPPAAL
intermediate format introduced in UPPAAL 3.6 Beta 1. 

Position tracking (the mechanism used to map an expression to some
position in the input file) has been reimplemented. The representation
of types in UTAP has been reimplemented. 

The interpreter for expressions has been removed. It was broken and
could not be fixed without reimplementing major parts of the UPPAAL
interpreter (not included in UTAP). This also means that errors
related to the range of variables or expressions are no longer
detected by the TypeChecker class.

New in 0.80

This is the version used in UPPAAL 3.6 Alpha 1. It contains a lot of
bug fixes since version 0.70. Likely there are some API
incompatibilities, but it should not be hard to adapt your program to
version 0.80. Notice that configure now accepts the --enable-cora
option to enable the extensions used by UPPAAL CORA.

New in 0.70
* This is a backport of the internal development code
* This is the parser used in Uppaal 3.4.8.
* Requires at least libxml2 2.6.10.
* The top level parser invocation functions like parseXTA now only
  run the type checker if the previous pass over the input did not
  produce any errors. This should make the parser much more robust.
* As always, we have fixed a lot of bugs.
* Most of the changes relate to the new syntax (which is used in 
  future versions of Uppaal).
* The most visible API change is that transition_t has been renamed 
  to edge_t.

New in 0.60
* This is a backport of the internal development code
* API changes include a split of SystemBuilder into several
  classes.
* Now requires at least libxml2 2.6.0.
* This is the parser used in Uppaal 3.4.6.
* Notice that header files now use a .h suffix.

New in 0.50
* Fixed bugs 32, 43, 46, 54, 58, 60, 82, 85, 88, 93, 97	and more
* Added tracer example program: Reads trace files in UPPAAL XTR format.
* Notice that some of the API has changed.

New in 0.40
* Bug fixes
* Reimplemented expression representation

New in 0.30
* Bug fixes
* Reimplementation of type and symbol handling, see the new
  symbol_t, type_t and frame_t classes.
* Reimplementation of TimedAutomataSystem construction and  
  type checking. The construction is now handled by the 
  SystemBuilder class and the type checking is done in a
  seperate pass by the new TypeChecker class.
* Range checking has been moved to another class: RangeChecker

New in 0.10
* Initial release


src/tags.cc
/* C++ code produced by gperf version 3.1 */
/* Command-line: gperf -C -E -t -L C++ -c -K str -Z Tags tags.gperf  */
/* Computed positions: -k'1,3' */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
#error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gperf@gnu.org>."
#endif

#line 1 "tags.gperf"
struct Tag { const char *str; tag_t tag; };
/* maximum key range = 48, duplicates = 0 */

class Tags
{
private:
  static inline unsigned int hash (const char *str, size_t len);
public:
  static const struct Tag *in_word_set (const char *str, size_t len);
};

inline unsigned int
Tags::hash (const char *str, size_t len)
{
  static const unsigned char asso_values[] =
    {
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 10, 15,  5,
      35, 10, 10,  5, 52,  0, 52, 52, 10,  5,
      20,  5, 20,  5, 20,  0, 15,  5, 52, 52,
      52,  0, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52, 52, 52, 52, 52,
      52, 52, 52, 52, 52, 52
    };
  return len + asso_values[static_cast<unsigned char>(str[2])] + asso_values[static_cast<unsigned char>(str[0])];
}

const struct Tag *
Tags::in_word_set (const char *str, size_t len)
{
  enum
    {
      TOTAL_KEYWORDS = 36,
      MIN_WORD_LENGTH = 3,
      MAX_WORD_LENGTH = 13,
      MIN_HASH_VALUE = 4,
      MAX_HASH_VALUE = 51
    };

  static const struct Tag wordlist[] =
    {
      {""}, {""}, {""}, {""},
#line 12 "tags.gperf"
      {"init",		TAG_INIT},
      {""},
#line 8 "tags.gperf"
      {"system",		TAG_SYSTEM},
      {""},
#line 28 "tags.gperf"
      {"instance",       TAG_INSTANCE},
#line 34 "tags.gperf"
      {"sbml",           TAG_SBML},
      {""},
#line 17 "tags.gperf"
      {"source",		TAG_SOURCE},
#line 30 "tags.gperf"
      {"message",        TAG_MESSAGE},
#line 7 "tags.gperf"
      {"instantiation",	TAG_INSTANTIATION},
#line 25 "tags.gperf"
      {"yloccoord",      TAG_YLOCCOORD},
      {""},
#line 14 "tags.gperf"
      {"urgent",		TAG_URGENT},
#line 38 "tags.gperf"
      {"comment",        TAG_COMMENT},
#line 22 "tags.gperf"
      {"lsc",            TAG_LSC},
#line 15 "tags.gperf"
      {"committed",	TAG_COMMITTED},
#line 36 "tags.gperf"
      {"query",          TAG_QUERY},
#line 33 "tags.gperf"
      {"anchor",         TAG_ANCHOR},
#line 35 "tags.gperf"
      {"queries",        TAG_QUERIES},
#line 11 "tags.gperf"
      {"location",	TAG_LOCATION},
#line 20 "tags.gperf"
      {"nail",		TAG_NAIL},
      {""},
#line 26 "tags.gperf"
      {"lsclocation",    TAG_LSCLOCATION},
#line 4 "tags.gperf"
      {"imports",	TAG_IMPORTS},
#line 6 "tags.gperf"
      {"template",	TAG_TEMPLATE},
#line 9 "tags.gperf"
      {"name",		TAG_NAME},
#line 19 "tags.gperf"
      {"label",		TAG_LABEL},
#line 29 "tags.gperf"
      {"temperature",    TAG_TEMPERATURE},
#line 21 "tags.gperf"
      {"project",        TAG_PROJECT},
#line 3 "tags.gperf"
      {"nta",		TAG_NTA},
#line 31 "tags.gperf"
      {"condition",      TAG_CONDITION},
#line 13 "tags.gperf"
      {"transition",	TAG_TRANSITION},
#line 16 "tags.gperf"
      {"branchpoint",    TAG_BRANCHPOINT},
#line 37 "tags.gperf"
      {"formula",        TAG_FORMULA},
#line 27 "tags.gperf"
      {"prechart",       TAG_PRECHART},
#line 23 "tags.gperf"
      {"type",           TAG_TYPE},
      {""},
#line 18 "tags.gperf"
      {"target",		TAG_TARGET},
      {""}, {""},
#line 24 "tags.gperf"
      {"mode",           TAG_MODE},
      {""},
#line 32 "tags.gperf"
      {"update",         TAG_UPDATE},
      {""}, {""},
#line 10 "tags.gperf"
      {"parameter",	TAG_PARAMETER},
      {""},
#line 5 "tags.gperf"
      {"declaration",	TAG_DECLARATION}
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      unsigned int key = hash (str, len);

      if (key <= MAX_HASH_VALUE)
        {
          const char *s = wordlist[key].str;

          if (*str == *s && !strncmp (str + 1, s + 1, len - 1) && s[len] == '\0')
            return &wordlist[key];
        }
    }
  return 0;
}
#line 39 "tags.gperf"

